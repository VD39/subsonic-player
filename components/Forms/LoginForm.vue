<script setup lang="ts">
import { config } from '@/config';
import { createForm, validateInputs } from './utils';
import MassageBar from '@/components/MessageBar/MessageBar.vue';
import InputField from '@/components/FormFields/InputField.vue';
import LoadingButton from '@/components/Buttons/LoadingButton.vue';

defineProps<{
  isLoading?: boolean;
  errorMessage?: string;
}>();

const emit = defineEmits(['submit']);

const formInputs = {
  server: {
    value: config.serverUrl,
    validationRules: {
      required: true,
      isUrl: true,
    },
  },
  username: {
    validationRules: {
      required: true,
    },
  },
  password: {
    validationRules: {
      required: true,
    },
  },
};

const form = createForm(formInputs);

async function submitForm() {
  validateInputs(form);

  if (!form.isValid.value) {
    return;
  }

  const { server, username, password } = form.fields;

  emit('submit', {
    password: password.value.value,
    server: server.value.value,
    username: username.value.value,
  });
}
</script>

<template>
  <form novalidate @submit.stop.prevent="submitForm">
    <div :class="$style.fields">
      <InputField
        v-if="!config.serverUrl"
        :id="form.fields.server.id"
        ref="serverUrl"
        v-model="form.fields.server.value.value"
        :class="$style.field"
        :label="form.fields.server.label"
        placeholder="Enter your server URL"
        required
        :has-error="!form.fields.server.isValid.value"
        :error-message="form.fields.server.errorMessage.value"
      />

      <InputField
        :id="form.fields.username.id"
        ref="username"
        v-model="form.fields.username.value.value"
        :class="$style.field"
        :label="form.fields.username.label"
        placeholder="Enter your username"
        required
        :has-error="!form.fields.username.isValid.value"
        :error-message="form.fields.username.errorMessage.value"
      />

      <InputField
        :id="form.fields.password.id"
        ref="password"
        v-model="form.fields.password.value.value"
        :class="$style.field"
        type="password"
        :label="form.fields.password.label"
        placeholder="Enter your password"
        required
        :has-error="!form.fields.password.isValid.value"
        :error-message="form.fields.password.errorMessage.value"
      />
    </div>

    <MassageBar v-if="errorMessage" :class="$style.messageBar" type="error">
      <p>{{ errorMessage }}</p>
    </MassageBar>

    <LoadingButton
      :class="$style.button"
      :disabled="isLoading"
      :is-loading="isLoading"
      is-full-width
    >
      Login
    </LoadingButton>
  </form>
</template>

<style module>
.fields,
.messageBar {
  margin-bottom: var(--space-32);
}

.field {
  margin-bottom: var(--space-12);
}
</style>
