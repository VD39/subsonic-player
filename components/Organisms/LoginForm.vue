<script setup lang="ts">
import InputField from '@/components/Atoms/InputField.vue';
import MessageBar from '@/components/Atoms/MessageBar.vue';
import SubmitButton from '@/components/Molecules/SubmitButton.vue';

defineProps<{
  error?: null | string;
  loading?: boolean;
}>();

const emit = defineEmits(['submit']);

const config = useRuntimeConfig();
const { SERVER_URL } = config.public;

const formInputs = {
  password: {
    validationRules: {
      required: true,
    },
  },
  server: {
    validationRules: !SERVER_URL
      ? {
          isUrl: true,
          required: true,
        }
      : {},
    value: SERVER_URL,
  },
  username: {
    validationRules: {
      required: true,
    },
  },
};

const form = createForm(formInputs);

async function submitForm() {
  validateInputs(form);

  if (!form.isValid.value) {
    return;
  }

  const { password, server, username } = form.fields;

  emit('submit', {
    password: password.value.value,
    server: server.value.value,
    username: username.value.value,
  });
}
</script>

<template>
  <form novalidate @submit.stop.prevent="submitForm">
    <div class="formFields">
      <InputField
        v-if="!SERVER_URL"
        :id="form.fields.server.id"
        ref="serverUrl"
        v-model="form.fields.server.value.value"
        class="formField"
        :label="form.fields.server.label"
        placeholder="Enter your server URL"
        required
        :error="form.fields.server.error.value"
      />

      <InputField
        :id="form.fields.username.id"
        ref="username"
        v-model="form.fields.username.value.value"
        class="formField"
        :label="form.fields.username.label"
        placeholder="Enter your username"
        required
        :error="form.fields.username.error.value"
      />

      <InputField
        :id="form.fields.password.id"
        ref="password"
        v-model="form.fields.password.value.value"
        class="formField"
        type="password"
        :label="form.fields.password.label"
        placeholder="Enter your password"
        required
        :error="form.fields.password.error.value"
      />
    </div>

    <MessageBar v-if="form.isValid.value && error" class="mBM" type="error">
      <p class="sentenceCase">{{ error }}</p>
    </MessageBar>

    <SubmitButton class="formButton" :loading="loading" full-width>
      Login
    </SubmitButton>
  </form>
</template>
