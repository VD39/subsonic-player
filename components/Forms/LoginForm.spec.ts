import type { VueWrapper } from '@vue/test-utils';
import { mount } from '@vue/test-utils';
import MassageBar from '@/components/MessageBar/MessageBar.vue';
import LoginForm from './LoginForm.vue';

const { config } = vi.hoisted(() => ({
  config: {
    serverUrl: '',
  },
}));

vi.mock('@/config', () => ({
  config,
}));

function factory(props = {}) {
  return mount(LoginForm, {
    props: {
      ...props,
    },
  });
}

describe('LoginForm', () => {
  let wrapper: VueWrapper;

  beforeEach(() => {
    wrapper = factory();
  });

  it('matches the snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  describe('when errorMessage prop is not set', () => {
    it('does not show the MassageBar component', () => {
      expect(wrapper.findComponent(MassageBar).exists()).toBe(false);
    });
  });

  describe('when errorMessage prop is set as true', () => {
    beforeEach(() => {
      wrapper = factory({
        errorMessage: 'Error message',
      });
    });

    it('matches the snapshot', () => {
      expect(wrapper.html()).toMatchSnapshot();
    });

    it('shows the MassageBar component', () => {
      expect(wrapper.findComponent(MassageBar).exists()).toBe(true);
    });
  });

  describe('when serverUrl in config is not defined', () => {
    it('shows the InputField component for server URL', () => {
      expect(wrapper.findComponent({ ref: 'serverUrl' }).exists()).toBe(true);
    });
  });

  describe('when serverUrl in config is defined', () => {
    beforeEach(() => {
      config.serverUrl = 'serverUrl';
      wrapper = factory();
    });

    it('matches the snapshot', () => {
      expect(wrapper.html()).toMatchSnapshot();
    });

    it('does not show the InputField component for server URL', () => {
      expect(wrapper.findComponent({ ref: 'serverUrl' }).exists()).toBe(false);
    });
  });

  describe('when form is invalid', () => {
    beforeEach(async () => {
      await wrapper.trigger('submit');
    });

    it('does not emit submit event', () => {
      expect(wrapper.emitted('submit')).toBe(undefined);
    });
  });

  describe('when form is valid', () => {
    beforeEach(async () => {
      config.serverUrl = '';
      wrapper = factory();

      wrapper
        .findComponent({ ref: 'serverUrl' })
        .vm.$emit('update:modelValue', 'https://www.test.com');
      wrapper
        .findComponent({ ref: 'username' })
        .vm.$emit('update:modelValue', 'username');
      wrapper
        .findComponent({ ref: 'password' })
        .vm.$emit('update:modelValue', 'password');
      await wrapper.trigger('submit');
    });

    it('matches the snapshot', () => {
      expect(wrapper.html()).toMatchSnapshot();
    });

    it('emits submit event with form values', () => {
      expect(wrapper.emitted('submit')).toEqual([
        [
          {
            password: 'password',
            server: 'https://www.test.com',
            username: 'username',
          },
        ],
      ]);
    });
  });
});
