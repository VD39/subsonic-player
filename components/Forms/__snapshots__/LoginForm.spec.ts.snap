// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LoginForm > matches the snapshot 1`] = `
"<form novalidate="">
  <div class="fields">
    <div class="inputField field"><label class="label" for="server">server <span class="required">*</span></label><input id="server" type="text" autocomplete="off" required="" placeholder="Enter your server URL" class="input" value="">
      <!--v-if-->
    </div>
    <div class="inputField field"><label class="label" for="username">username <span class="required">*</span></label><input id="username" type="text" autocomplete="off" required="" placeholder="Enter your username" class="input" value="">
      <!--v-if-->
    </div>
    <div class="inputField field"><label class="label" for="password">password <span class="required">*</span></label><input id="password" type="password" autocomplete="off" required="" placeholder="Enter your password" class="input" value="">
      <!--v-if-->
    </div>
  </div>
  <!--v-if--><button class="submitButton fullWidth button" type="submit"><span class="content"> Login <!--v-if--></span></button>
</form>"
`;

exports[`LoginForm > when errorMessage prop is set as true > matches the snapshot 1`] = `
"<form novalidate="">
  <div class="fields">
    <div class="inputField field"><label class="label" for="server">server <span class="required">*</span></label><input id="server" type="text" autocomplete="off" required="" placeholder="Enter your server URL" class="input" value="">
      <!--v-if-->
    </div>
    <div class="inputField field"><label class="label" for="username">username <span class="required">*</span></label><input id="username" type="text" autocomplete="off" required="" placeholder="Enter your username" class="input" value="">
      <!--v-if-->
    </div>
    <div class="inputField field"><label class="label" for="password">password <span class="required">*</span></label><input id="password" type="password" autocomplete="off" required="" placeholder="Enter your password" class="input" value="">
      <!--v-if-->
    </div>
  </div>
  <div class="error messageBar messageBar">
    <p>Error message</p>
  </div><button class="submitButton fullWidth button" type="submit"><span class="content"> Login <!--v-if--></span></button>
</form>"
`;

exports[`LoginForm > when form is valid > matches the snapshot 1`] = `
"<form novalidate="">
  <div class="fields">
    <div class="inputField field"><label class="label" for="server">server <span class="required">*</span></label><input id="server" type="text" autocomplete="off" required="" placeholder="Enter your server URL" class="input" value="https://www.test.com">
      <!--v-if-->
    </div>
    <div class="inputField field"><label class="label" for="username">username <span class="required">*</span></label><input id="username" type="text" autocomplete="off" required="" placeholder="Enter your username" class="input" value="username">
      <!--v-if-->
    </div>
    <div class="inputField field"><label class="label" for="password">password <span class="required">*</span></label><input id="password" type="password" autocomplete="off" required="" placeholder="Enter your password" class="input" value="password">
      <!--v-if-->
    </div>
  </div>
  <!--v-if--><button class="submitButton fullWidth button" type="submit"><span class="content"> Login <!--v-if--></span></button>
</form>"
`;

exports[`LoginForm > when serverUrl in config is defined > matches the snapshot 1`] = `
"<form novalidate="">
  <div class="fields">
    <!--v-if-->
    <div class="inputField field"><label class="label" for="username">username <span class="required">*</span></label><input id="username" type="text" autocomplete="off" required="" placeholder="Enter your username" class="input" value="">
      <!--v-if-->
    </div>
    <div class="inputField field"><label class="label" for="password">password <span class="required">*</span></label><input id="password" type="password" autocomplete="off" required="" placeholder="Enter your password" class="input" value="">
      <!--v-if-->
    </div>
  </div>
  <!--v-if--><button class="submitButton fullWidth button" type="submit"><span class="content"> Login <!--v-if--></span></button>
</form>"
`;
